src/ast/ast.h

```diff
@@ -61,22 +61,49 @@ enum class ExpansionType {
 
 class Node {
 public:
-  Node(location loc) : loc(loc) {};
+  Node(Diagnostics &d, location loc) : diagnostics_(d), loc(loc) {};
   virtual ~Node() = default;
 
   Node(const Node &) = delete;
   Node &operator=(const Node &) = delete;
   Node(Node &&) = delete;
   Node &operator=(Node &&) = delete;
 
+  template <typename... Args>
+  Diagnostic &addError(Args &...args) const
+  {
+    if constexpr (sizeof...(Args) == 0) {
+      return diagnostics_.addError(loc);
+    } else {
+      return diagnostics_.addError(loc + (args.loc + ...));
```

Comment:
This looks neat (automatically adding the location). Help my sad little brain understand the else branch - you add the node's location and then an additional args location? Mind pointing me to an example of this?

Reply:
It just automatically merges all the locations for any additional nodes that you pass. For example,

```
foo.addError(bar)
```

I just made this require a second node to keep a consistent API (and allow for future metadata, if we prefer to do that via some diagnostic structure instead of the location). I believe that there are one or two instances in the next change, which is why this was added. See `binop_ptr` in the semantic analyser, e.g. `binop.left->addWarning(*binop.right)`